File: leetcode/Python/0113-m-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

"""
Recursion - traversal
Time: O(n)
Space: O(n)
"""
class Solution_1:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if not root:
            return []
        res = []
        self.helper(root, sum, [], res)
        return res
    
    def helper(self, root, sum, path, res):
        if not root.left and not root.right and sum == root.val:
            path.append(root.val)
            res.append(path)
        if root.left:
            self.helper(root.left, sum - root.val, path + [root.val], res)
        if root.right:
            self.helper(root.right, sum - root.val, path + [root.val], res)

"""
BFS
Time: O(n)
Space: O(n)
"""
class Solution_2:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        if not root:
            return []
        res = []
        queue = [(root, root.val, [root.val])]
        while queue:
            curr, val, ls = queue.pop(0)
            if not curr.left and not curr.right and val == sum:
                res.append(ls)
            if curr.left:
                queue.append((curr.left, val + curr.left.val, ls + [curr.left.val]))
            if curr.right:
                queue.append((curr.right, val + curr.right.val, ls + [curr.right.val]))
        return res

File: leetcode/C++/0113-m-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// Time: O(N^2)
// Space: O(N)
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> res;
        if (!root) return res;
        helper(root, sum, vector<int>(), res);
        return res;
    }
    
    void helper(TreeNode* root, int sum, vector<int> path, vector<vector<int>>& res) {
        if (!root) return;
        if (root->val == sum && !root->left && !root->right) {
            path.push_back(root->val);
            res.push_back(path);
            return;
        }
        sum -= root->val;
        path.push_back(root->val);
        helper(root->left, sum, path, res);
        helper(root->right, sum, path, res);
    }
};